# Profitable App Profiles for the App Store and Google Play Markets

This project will show which Apps are the most profitable from the App Store and Google Play Markets. 

The goal is to provide data driven reserch for Mobile App Developers to show where the market is trending. More specifically, we will be analyzing the free apps, and which are more profitable.

# Opening and Exploring the Data

Below the data is read from:
- (Google) https://www.kaggle.com/lava18/google-play-store-apps/home
- (Apple) https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home

These two datasets are opened and read into the the two variables below.

In [1]:
from csv import reader

def read_data(file, header = True):
    opened_file = open(file)
    read_file = reader(opened_file)
    dataset = list(read_file)
    if header:
        return dataset[0], dataset[1:]
    else:
        return dataset

# Apple Data Set
apple_header, apple_data = read_data('AppleStore.csv')
# Google Data Set
google_header, google_data = read_data('googleplaystore.csv')

The below function is used to print the rows in the dataset at a specified range. This will be used to explore that contents of the data above.

Now we will explore the Apple Data set by first looking a the header row and the amount of Columns and Rows. This is get and idea of the amount of data we are working with.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n')

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
print(apple_header,'\n')
explore_data(apple_data, 0, 3, True)        

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


The above shows that 7197 rows and 16 columns to work with. At a glance, some of the columns that might provide some value are "track_name", "price". "rating_count_tot", "user_rating", and "prime_genre". This will become more clear as we look into the data.
Since there are over 7000 rows, only first three rows are displayed.

Next we will look at a sample of the Google Data.

In [3]:
print(google_header,'\n')
explore_data(google_data, 0, 3, True)   

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The above shows that there are 10841 rows with 13 columns. Of the 13 columns, the ones that appear to be useful for the analytics are "App", "Category", "Rating", "Installs", "Price", "Rating", and "Genres". Again this will become more clear as we analyze the data.

# Deleting Wrong Data

After browsing the discussions pags on Kaggle, it became apparent that there is an error within one of the rows in the Google dataset. The Category is aparently missing, which will cause some issues in our analytics.

Please see https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015 for further details.

We will now look at the problem row and determine if it needs to be removed. This will be compared against a correct row for analysis.

In [4]:
print(google_header,'\n') #header
print(google_data[10472],'\n') #incorrect row
print(google_data[1]) #correct row

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


The above shows that row 10472 is indeed incorrect. The category appears to be completely missing with 1.9 in its place. This row will be delete from the data set for accuracy.

In [5]:
del google_data[10472]

# Removing Duplicate Entries
## Part One
To ensure the data sets we are working with are clean and valid, we must verify that there no duplicate entries. This can be done by first finding any two apps with the same name.

In [6]:
duplicate_apps = []
unique_apps = []

for row in google_data:
    name = row[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('The number of duplicate apps is:', len(duplicate_apps),'\n')
print('Examples of duplicate apps:', duplicate_apps[:10])

The number of duplicate apps is: 1181 

Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


As you can see above, there are a total of 1181 duplicate apps in the data set. The next question is which of the duplicates do we keep? For this, I will propose that we keep the rows with the latest date in the "Last Updated" column. Once the data is clean, we should have the following number of clean rows.

## Part Two
For this section, we will create a dictionary to store the unique rows with the greatest "Last Updated" date. We will need to use the datetime module to compare the dates. To keep the data as close to its original state as possible, we will only use the module for comparison. The actual string value will still be stored in the dictionary.

In [7]:
from datetime import datetime
dates_max = {}

for row in google_data:
    name = row[0]
    last_updated = row[10]
    last_updated_object = datetime.strptime(last_updated, "%B %d, %Y")
    if name in dates_max:
        if datetime.strptime(dates_max[name], "%B %d, %Y") < last_updated_object:
            dates_max[name] = last_updated
    if name not in dates_max:
        dates_max[name] = last_updated
        
print('Expected Rows:',len(google_data) - len(duplicate_apps))
print('Actual Rows:',len(dates_max))

Expected Rows: 9659
Actual Rows: 9659


As you can see above, we are succesful in keeping the most up to date apps in our disctionary. The next step is to use this disctionary to remove the duplicate rows in our actual dataset.

In [8]:
google_data_clean = []
already_added = []

for row in google_data:
    name = row[0]
    last_updated = row[10]
    if (dates_max[name] == last_updated) and (name not in already_added):
        google_data_clean.append(row)
        already_added.append(name)

Now we will explore the new clean data set to verify the duplicates were removed.

In [9]:
explore_data(google_data_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9659
Number of columns: 13


From the sample above, this looks to be correct. We now have a Google dataset with no duplicates to work with.

# Removing Non-English Apps
## Part One

Since we are only interested in the apps that are in English, the ones that use anything other than English Characters in their title will be removed.

The first step here is to create a function that will read the title of the app and determine if it has Non-English characters.

In [10]:
def is_english(string):
    for char in string:
        if ord(char) > 127:
            return False
    return True    

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


The function is working, but it is also returning false for charcters such as emoji's or symbols. This will need to be corrected to avoid removing English apps.

## Part Two

The function needs to be re-written to accept English apps with special charcters in it. We can do these by returning false only if the name contains more than three Non-English characters.

In [11]:
def is_english(string):
    non_ascii = 0
    
    for char in string:
        if ord(char) > 127:
            non_ascii += 1

    if non_ascii > 3:
        return False
    else:
        return True    

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


The above function will now work to determine which apps are English, including those with special characters. This function will now be used to filter the Non-English apps in both the Apple and Google datasets.

In [12]:
apple_english = []
google_english = []

for row in apple_data:
    name = row[1]
    if is_english(name) == True:
        apple_english.append(row)
        
for row in google_data_clean:
    name = row[0]
    if is_english(name) == True:
        google_english.append(row)
        
explore_data(apple_english, 0, 3, True)
print('\n')
explore_data(google_english, 0, 3, True)        

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1,

We can see that we're left with 9614 Android apps and 6183 iOS apps.

# Isolating the Free Apps
Since we stated in the beginning that we are interested in the free apps, we need to isolate those into their own dataset.

In [13]:
google_free = []
apple_free = []

for row in google_english:
    price = row[7]
    if price == '0':
        google_free.append(row)
        
for row in apple_english:
    price = row[4]
    if price == '0.0':
        apple_free.append(row)
        
print(len(google_free))
print(len(apple_free))

8864
3222


The above has taken a decent chunk out of our datasets. Our free apps are now isolated and should be ready for analysis.

# Most Common Apps by Genre
## Part One

The rest of this analysis will be done by review the most profitable apps by genre. Genre will be a good key indicator of what type of app developers should be creating. 

This analysis will be done by creating frequency tables to determine which app genre is most popular in both markets. Both datasets have genre columns, which will be very useful when conducting the analysis.

## Part Two
Well will begin by creating two functions to generate the frequency tables. 
- The first function will be used to create and return the frequency tables
- The second will be used to display and sort the table

In [14]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for row in dataset:
        total += 1
        genre = row[index]
        if genre in table:
            table[genre] += 1
        else:
            table[genre] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

## Part Three
We will look at the Google Dataset first. Keep in mind this dataset only contains Free English apps.
The Google Dataset contains both a Category and Genre Column. We will explore both, but Category first.

In [15]:
display_table(google_free, 1) #Category

FAMILY : 18.479241877256317
GAME : 9.87138989169675
TOOLS : 8.461191335740072
BUSINESS : 4.580324909747293
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2490974729241873
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.782490974729242
MAPS_AND_NAVIGATION : 1.3989169675090252
EDUCATION : 1.2522563176895307
FOOD_AND_DRINK : 1.2409747292418771
ENTERTAINMENT : 1.128158844765343
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
ART_AND_DESIGN : 0.6768953068592057
PARENTING : 0.6

The above shows that Family is the most common Category, with Games and Tools being nearly half as popular. Family and Games would be geared more towards entertainment. 
We will see if this still holds true when explore the Genre column.

In [16]:
display_table(google_free, 9) #Genre

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.580324909747293
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2490974729241873
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

This is a much larger list, which I would argue is a more broken out version of the Categories.
The top Genres here are Tools, Entertainment, and Education. This a nice variety of Entertainment and Practicality.

We will now explore the Apple dataset to see how it compares.

In [17]:
display_table(apple_free, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see above that Apple App Store has a huge audience for Games. In contrast, the Google Play Store was a little more diverse, but Games and Entertainment were still popular. This paints a somewhat more clear picture of what's dominating the market.

# Most Popular Apps by Genre on the App Store
To help paint a better picture, we must look at the installs for each genre. More specifically, we will begin by looking at the average installs for each Genre.

We will start with the Apple DataSet.

In [18]:
genres_ios = freq_table(apple_free, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for row in apple_free:
        genre_app = row[-5]
        if genre_app == genre:
            user_ratings = float(row[5])
            total += user_ratings
            len_genre += 1
    avg_ratings = total / len_genre
    print(genre, avg_ratings)

News 21248.023255813954
Food & Drink 33333.92307692308
Productivity 21028.410714285714
Lifestyle 16485.764705882353
Entertainment 14029.830708661417
Weather 52279.892857142855
Sports 23008.898550724636
Music 57326.530303030304
Travel 28243.8
Utilities 18684.456790123455
Shopping 26919.690476190477
Reference 74942.11111111111
Education 7003.983050847458
Photo & Video 28441.54375
Games 22788.6696905016
Medical 612.0
Navigation 86090.33333333333
Finance 31467.944444444445
Book 39758.5
Business 7491.117647058823
Social Networking 71548.34905660378
Catalogs 4004.0
Health & Fitness 23298.015384615384


On average, it appears that Navigation apps have the highest average user ratings. While Games appeard to take a up a large portion of the App Store, they are not rated nearly as high. Reference is another genre with a high user rating. These include books (The Bible or a Dictionary) in the form of an app.

# Most Popular Apps by Genre on Google Play

We will now conduct a similar analysis on the data for the Google Playstore.

However, the Google Dataset actually tells us the number of installs. This will be used instead as it gives us better information about the app.

In [20]:
genres_google = freq_table(google_free, 1)

for genre in genres_google:
    total = 0
    len_category = 0
    for row in google_free:
        app_genre = row[1]
        if app_genre == genre:
            installs = row[5]
            installs = installs.replace("+","")
            installs = float(installs.replace(",",""))
            total += installs
            len_category += 1
    avg_installs = total / len_category
    print(genre, avg_installs)

WEATHER 5074486.197183099
NEWS_AND_MAGAZINES 9549178.467741935
PERSONALIZATION 5201482.6122448975
ART_AND_DESIGN 1905351.6666666667
BOOKS_AND_REFERENCE 8767811.894736841
FOOD_AND_DRINK 1924897.7363636363
COMMUNICATION 38326063.197916664
SOCIAL 23253652.127118643
FINANCE 1387692.475609756
VIDEO_PLAYERS 24852732.40506329
PHOTOGRAPHY 17840110.40229885
TOOLS 10801391.298666667
MEDICAL 120263.07987220447
LIBRARIES_AND_DEMO 638503.734939759
MAPS_AND_NAVIGATION 4056941.7741935486
COMICS 817657.2727272727
EVENTS 253542.22222222222
BUSINESS 1704192.3399014778
TRAVEL_AND_LOCAL 13984077.710144928
SPORTS 3638640.1428571427
PARENTING 542603.6206896552
HOUSE_AND_HOME 1331540.5616438356
GAME 15826136.514285713
PRODUCTIVITY 16787331.344927534
HEALTH_AND_FITNESS 4188821.9853479853
EDUCATION 2169819.8198198196
ENTERTAINMENT 21134600.0
FAMILY 2814893.583638584
BEAUTY 513151.88679245283
SHOPPING 7036877.311557789
AUTO_AND_VEHICLES 647317.8170731707
LIFESTYLE 1437816.2687861272
DATING 854028.8303030303


The above shows Communication is dominating the market by having 38,326,063 installs. Entertainment is also huge with installs of 21134600. 

# Conclusion

While Games and Entertainment seem to be huge in both markets, their ratings aren't nearly as high as their amount of installs. Creating a free game for both platforms will most likely get lost in the sea of other free games. 

An interesting idea would actually be to make a free Reference app in both markets. They are rated highly and both very popular. The development time and cost would also be pretty low since the app is most text without much logic.